### PR TITLE
ResamplingSpCVBuffer

### LINKED ISSUE #13: Support presence-background option in "Spatial Buffer CV"
If the target has a binary outcome, a presence-background approach (see blockCV::buffering) would be possible. Target needs to be transformed to 0/1 before sampling.

### PR DESCRIPTION
**Pull request by be-marc**
Closes #13 

* Enables spDataType and addBG parameter
* Adds parameter tests
* Fixes train and test set storage
* Adds 6x6 point grid tasks (two-class, multi-class and continuous response)
* Adds unit tests for `ResamplingSpCVBuffer` with 6x6 point grid tasks
* Update autoplot unit test `vdiffr` images

This implementation uses the parameter names of `blockCV::buffering` which makes it easier to call the function (via `mlr3misc::invoke`)  with its default parameter values. However, this leaves us with the rather ugly parameter name `theRange`. If we stick to the parameter names of `blockCV` we should also change `ResamplingSpCVBlock`. Maybe we should discus this once again.

### CODE REVIEW

**Comment by pat-s** [comment]
> If we stick to the parameter names of blockCV we should also change ResamplingSpCVBlock. Maybe we should discus this once again.

Hm yeah. We probably should because this makes it easier for users.


---

@be-marc I am a bit confused about the order of the PRs. They all contain overlapping content. Which one should go first?

**Comment by be-marc** [comment]
> Hm yeah. We probably should because this makes it easier for users.

I will do this after the merge of the three PRs

ResamplingSpCVBuffer -> Fix *Env classes -> Unit tests

File: R/ResamplingSpCVBuffer.R [hunk 1]
```diff
@@ -5,6 +5,17 @@
...
+#' The `blockcv::buffering` function was written for species distribution modeling with presence-absence data.
+#' However, the default settings allow to conduct a leave-one-out cross validation
+#' for two-class, multi-class and continuous response data, where each observation is one test set.
+#' For each test, all observations outside the buffer around the test observation are included in the training set.
+#'
+#' The parameter `spDataType = PB` and `addBG` are designed for presence-background data in species distribution modelling.
+#' If `spDataType = PB`, test sets are only created for each presence observation (`task\$postive`).
+#' The option `addBG = TRUE` adds the background data inside the buffer to the corresponding test sets.
+#' For each test set, all observations outside the buffer around the test observation are included in the training set.
+#'
```
Comment (pat-s) [inline R/ResamplingSpCVBuffer.R]: chr width of 80

File: R/ResamplingSpCVBuffer.R [hunk 2]
```diff
@@ -72,47 +89,55 @@ ResamplingSpCVBuffer = R6Class("ResamplingSpCVBuffer",
...
   private = list(
-    .sample = function(ids, coords, crs) {
+    .sample = function(ids, response, coords, positive, crs, properties) {
 
       require_namespaces(c("blockCV", "sf"))
 
-      points = sf::st_as_sf(coords,
+      pars = self$param_set$get_values()
+
+      if (!isTRUE("twoclass" %in% properties) && isTRUE(pars$spDataType == "PB")) {
```
Comment (pat-s) [inline R/ResamplingSpCVBuffer.R]: Please remove `isTRUE()` calls.

File: R/ResamplingSpCVBuffer.R [hunk 3]
```diff
@@ -72,47 +89,55 @@ ResamplingSpCVBuffer = R6Class("ResamplingSpCVBuffer",
...
+        stop("spDataType = 'PB' should only be used with two-class response")
+      }
+
+      if (!is.null(pars$addBG) && isTRUE(pars$spDataType == "PA")) {
+        stop("Parameter addBG should only be used with spDataType = 'PB'")
```
Comment (pat-s) [inline R/ResamplingSpCVBuffer.R]: `stopf()`

File: R/ResamplingSpCVBuffer.R [hunk 4]
```diff
@@ -72,47 +89,55 @@ ResamplingSpCVBuffer = R6Class("ResamplingSpCVBuffer",
...
-        theRange = self$param_set$values$range,
-        progress = FALSE
-      )
+      inds = invoke(blockCV::buffering,
+        speciesData = data,
+        progress = FALSE,
+        .args = pars)
 
-      inds = map(inds$folds, function(x) {
+      map(inds$folds, function(x) {
```
Comment (pat-s) [inline R/ResamplingSpCVBuffer.R]: `mlr3misc::map()`

File: R/autoplot.R [hunk 5]
```diff
@@ -123,11 +123,91 @@ autoplot.ResamplingSpCVBuffer = function(
...
+  coords$row_id = task$row_ids
+  require_namespaces(c("sf", "cowplot"))
+
+  # instantiate if not yet done
+  if (!resampling$is_instantiated) {
+    resampling = resampling$instantiate(task)
+  }
+
+  if (is.null(fold_id)) {
+    stop("Plotting all folds of a LOOCV instance is not supported. Please provide a fold ID.") # nolint
```
Comment (pat-s) [inline R/autoplot.R]: `stopf()` and linebreak :)

File: R/autoplot.R [hunk 6]
```diff
@@ -123,11 +123,91 @@ autoplot.ResamplingSpCVBuffer = function(
...
+  }
+
+  # plot train and test of a specific fold?
+  if (length(fold_id) > resampling$iters) {
+    stop("More folds specified than stored in resampling.")
```
Comment (pat-s) [inline R/autoplot.R]: stopf()

File: R/autoplot.R [hunk 7]
```diff
@@ -123,11 +123,91 @@ autoplot.ResamplingSpCVBuffer = function(
...
+  }
+  if (length(fold_id) == 1 && fold_id > resampling$iters) {
+    stop("Specified a fold id which exceeds the total number of folds.")
+  }
+  if (any(fold_id > resampling$iters)) {
+    stop("Specified a fold id which exceeds the total number of folds.")
```
Comment (pat-s) [inline R/autoplot.R]: stopf()

File: R/autoplot.R [hunk 8]
```diff
@@ -123,11 +123,91 @@ autoplot.ResamplingSpCVBuffer = function(
...
+  if (is.null(fold_id)) {
+    stop("Plotting all folds of a LOOCV instance is not supported. Please provide a fold ID.") # nolint
+  }
+
+  # plot train and test of a specific fold?
+  if (length(fold_id) > resampling$iters) {
+    stop("More folds specified than stored in resampling.")
+  }
+  if (length(fold_id) == 1 && fold_id > resampling$iters) {
+    stop("Specified a fold id which exceeds the total number of folds.")
```
Comment (pat-s) [inline R/autoplot.R]: stopf()

File: R/ResamplingSpCVBuffer.R (hunk 2 above)
Comment (be-marc) [reply R/ResamplingSpCVBuffer.R]: We can remove the first `isTRUE` but the second condition produces an error without `isTRUE` if `pars$spDataType` is not set (We would use the package default in this case). If you want we can set `pars$spDataType == "PA"` before checking the condition if `pars$spDataType` is not set.

**Comment by be-marc** [comment]
I merged the changes from the other PR.

File: R/ResamplingSpCVBuffer.R (hunk 2 above)
Comment (pat-s) [reply R/ResamplingSpCVBuffer.R]: Hm I see. Maybe we should just leave it and `isTRUE` is not as bad as I think about it 😄

**Pull request merged**